java/core/src/main/java/com/microsoft/bond/BondType.java

```diff
@@ -0,0 +1,501 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package com.microsoft.bond;
+
+import com.microsoft.bond.helpers.ArgumentHelper;
+import com.microsoft.bond.protocol.ProtocolWriter;
+import com.microsoft.bond.protocol.TaggedProtocolReader;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Represents a type in the Bond type system and implements behavior specific to that type, such as
+ * schema inference, initialization, serialization and deserialization. A Bond type cane be a primitive,
+ * a struct, a specialization of a generic type (including containers, nullable, or generic Bond structs).
+ * @param <T> the class of the value
+ */
+public abstract class BondType<T> {
+
+    // The global type cache
+    static final TypeCache typeCache = new TypeCache();
+
+    // Registry of struct type resolvers; entries are added by static class initializers
+    private static final ConcurrentHashMap<
+            Class<? extends BondSerializable>,
+            StructBondTypeResolver<? extends BondSerializable>> structTypeResolverRegistry =
+            new ConcurrentHashMap<
+                    Class<? extends BondSerializable>,
+                    StructBondTypeResolver<? extends BondSerializable>>();
+
+    // Registry of enum types; entries are added by static class initializers
```

Comment:
See above.

---
In reply to: [123089182](https://github.com/Microsoft/bond/pull/502#discussion_r123089182) [](ancestors = 123089182)